Rewrote `Dispatcher`

djspiewak opened the pull request:
I have fixed many things. I possibly also made new problems. I think it's more maintainable now, but who am I to say?

I wrote a long comment in the middle which kind of explains the implementation, but the tldr is we're now very explicit about the impure submission queue and its associated race conditions. Those race conditions are fully resolved out by `Worker`, including the cancelation state machine. On the "actually do stuff" side is `Executor`, which is quite simple for the existing modes and quite complex for the new mode: `sequentialCancelable`.

Speaking of which, I added a third mode because I didn't realize it wasn't already the semantics of the second mode (`sequential`). Whoops. Anyway, we might not want to expose this one since I'm not sure anyone wants it, but I did all the work so here you go. Mostly I wanted to prove that it was possible. And it is… just hard.

Oh, I also removed and rewrote a bunch of tests which either weren't right, weren't reliable, or weren't interesting. I added a few new tests covering cases we probably should have always had. I don't *think* I missed anything, but who can say? Please enjoy.

djspiewak committed:
Reimplemented Dispatcher.sequential(false)

djspiewak committed:
Added support for `Dispatcher.sequential(await = true)`

djspiewak committed:
Attempting to get `parallel` working

djspiewak committed:
Adjusted flaky and/or definitionally-wrong tests

djspiewak committed:
Fixed late cancelation with `sequential`

djspiewak committed:
Trying really hard to make it work...

djspiewak committed:
Reimplemented part of Dispatcher some more

djspiewak committed:
Made cancelable sequential accessible and fixed `cede` loop

djspiewak committed:
Rewrote cancelation state machine to be more explicit

djspiewak committed:
Renamed to avoid insanity

djspiewak committed:
Fixed sequential multi-cancelation

djspiewak committed:
Fixed fun memory race (on ARM no less!) with parallel cancelation

djspiewak committed:
Removed bad idea

djspiewak committed:
Made new `Dispatcher` mode package-private until we can decide what to do with it

djspiewak committed:
Replaced `LinkedBlockingQueue` with `UnsafeUnbounded`

djspiewak committed:
Scalafmt

djspiewak committed:
Fixed compilation on Scala 2.12 and 3.x

djspiewak committed:
Filtered out spurious mima warnings

djspiewak committed:
Skip hammering on race condition on non-JVM platforms

djspiewak committed:
V8 *really* doesn't like this test…

djspiewak commented:
Hard for me to tell if this is a legit failure or not, but there's at least *one* legit bug here somewhere. Got this locally just now:

```
[error]     ! not hang when cancelling
[error]      cats.effect.TestTimeoutException: null (Runners.scala:109)
[error] cats.effect.Runners.$anonfun$timeout$1(Runners.scala:109)
[error] cats.effect.unsafe.WorkStealingThreadPool.$anonfun$sleep$1(WorkStealingThreadPool.scala:685)
[error] cats.effect.unsafe.WorkStealingThreadPool.$anonfun$sleep$1$adapted(WorkStealingThreadPool.scala:685)
[error] cats.effect.unsafe.WorkerThread.run(WorkerThread.scala:740)
```

JVM, 2.12.18, ARM64 (parallel, await = false)

durban commented on std/shared/src/main/scala/cats/effect/std/Dispatcher.scala:
```diff
@@ -190,240 +196,444 @@ object Dispatcher {
+   * at the cost of ordering, which is not guaranteed in parallel mode. With the sequential modes, there
+   * is only a single worker.
+   *
+   * On the impure side, the queue bit is the easy part: it's just a `LinkedBlockingQueue` which
```
The LinkedBlockingQueue have been already replaced by UnsafeUnbounded, but this comment still mentions it.

djspiewak replied:
Good callout thank you!

djspiewak committed:
Scalafmt

djspiewak committed:
Fixed outdated comment

djspiewak committed:
Fixed missing case in `unsafeOffer` cas loop

samspills committed:
test ordering maintained during finalization

djspiewak committed:
Revert "test ordering maintained during finalization"

This reverts commit ecd9008d0f037bf1a7bcfcb14cd7c0587cbb3428.

djspiewak committed:
Added tests to assert queue draining behavior during shutdown

samspills committed:
test blocking cancellation

armanbilge committed:
Move blocked-cancelation test to parallel only

armanbilge committed:
Separate blocked-cancelation into two tests

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/DispatcherSpec.scala:
```diff
@@ -201,37 +310,67 @@ class DispatcherSpec extends BaseSpec with DetectPlatform {
+        }
+        .replicateA_(if (isJVM) 1000 else 1)
+        .as(ok)
+    }
```
This test verifies that the worker doesn't get blocked here:

https://github.com/typelevel/cats-effect/blob/abf680d64527489b41067b51224ca2629eedc21d/std/shared/src/main/scala/cats/effect/std/Dispatcher.scala#L449-L450

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/DispatcherSpec.scala:
```diff
@@ -201,37 +310,67 @@ class DispatcherSpec extends BaseSpec with DetectPlatform {
+        }
+        .replicateA_(if (isJVM) 1000 else 1)
+        .as(ok)
     }
```
This test verifies that the worker doesn't get blocked here.
https://github.com/typelevel/cats-effect/blob/abf680d64527489b41067b51224ca2629eedc21d/std/shared/src/main/scala/cats/effect/std/Dispatcher.scala#L427-L429

armanbilge commented on std/shared/src/main/scala/cats/effect/std/Dispatcher.scala:
```diff
@@ -190,240 +197,474 @@ object Dispatcher {
+        def apply(task: F[Unit])(registerCancel: F[Unit] => F[Unit]): F[Unit] = {
+          // we can use unit as a cancel action here since it must always sequence *after* the task
+          // thus, the task must complete before the cancel action will be picked up
+          registerCancel(Applicative[F].unit) *> task
```
```suggestion
          // we can use unit as a cancel action here since it must always sequence *after* the task
          // thus, the task must complete before the cancel action will be picked up and we do not need to backpressure on its termination as is expected of cancel
          registerCancel(Applicative[F].unit) *> task
```

armanbilge committed:
Implement `UnsafeAsyncQueue#take` with `Async#cont`

armanbilge committed:
Simplify `UnsafeAsyncQueue`

armanbilge committed:
Optimize `UnsafeAsyncQueue`

djspiewak committed:
Fixed new cancelation tests and fixed functionality

djspiewak commented:
Fixed all the things, I think. The first test was more deterministically expressible via `TestControl`. The second test was a deadlock even with corrected functionality, so I fixed that as well. Both tests rely on the latch release running on the dispatcher, which implies the workers are unblocked.

I am now seeing a rather odd issue though, which is occasionally something like this prints:

```
[info]     + cancelation does not block a worker
cats.effect.std.FailureSignal$
[info]     + cancelation race does not block a worker
[info]     + cancel all inner effects when canceled
[info] Total for specification DispatcherSpec
```

This suggests that `UnsafeUnbounded` is being taken while empty and the error is *not* handled, so it's falling through to the fiber default handler. This isn't ideal, obviously, and might be related to @armanbilge's changes.

armanbilge commented:
Yes, that was my fault, sorry! I've got a fix locally, will push it up a bit later.

samspills commented on std/shared/src/main/scala/cats/effect/std/Dispatcher.scala:
```diff
@@ -190,240 +198,456 @@ object Dispatcher {
+   * a submission queue with impure enqueue and cancel functions which is drained by the `Worker` and an
+   * internal execution protocol which also involves a queue. The `Worker` encapsulates all of the
+   * race conditions and negotiations with impure code, while the `Executor` manages running the
+   * tasks with appropriate semantics. In parallel mode, we shard the `Worker`s according to the
+   * number of CPUs and select a random queue (in impure code) as a target. This reduces contention
```
I'm not sure I understand this comment. In the existing (v3.5.3) implementation, a dispatcher has N workers and each worker has N queues (where N = number of CPUs). But in the rewrite there are N workers and each worker has just 1 queue (so we've lost a degree of sharding). I don't see where we're selecting a random queue, just where we select the random worker. Maybe this comment could be clarified? 

And then relatedly: do you have a feeling for the practical consequences of the reduced sharding? Is it okay to just shard across workers?

djspiewak replied:
This is a good question. I actually don't know. This suggests that we really do need `Dispatcher` benchmarks and I should stop putting it off. My intuition about the cost of the reduced sharding is that it's likely to be minimal. It's basically going to put a bit more load on the enqueueing side (in the form of contention) while simplifying and taking load off of the dequeue side (the worker).

durban commented on std/shared/src/main/scala/cats/effect/std/Dispatcher.scala:
```diff
@@ -190,240 +198,456 @@ object Dispatcher {
+          workersF evalMap { workers =>
+            Async[F].executionContext flatMap { ec =>
+              val launchAll = 0.until(workers.length).toList traverse_ { i =>
+                supervisor.supervise(workers(i).run).void
```
```suggestion
                supervisor.supervise(workers(i).run)
```

durban commented on std/shared/src/main/scala/cats/effect/std/Dispatcher.scala:
```diff
@@ -190,240 +198,456 @@ object Dispatcher {
+                      val reg = new Registration.Primary(promisory.void, stateR)
+                      worker.queue.unsafeOffer(reg)
+
+                      def cancel(): Future[Unit] = {
```
```suggestion
                      @tailrec
                      def cancel(): Future[Unit] = {
```

durban committed:
Add (sometimes) failing test

durban committed:
Try to fix the bug

durban commented on std/shared/src/main/scala/cats/effect/std/Dispatcher.scala:
```diff
@@ -190,240 +198,456 @@ object Dispatcher {
+                    // forward atomicity guarantees onto promise completion
+                    val promisory = MonadCancel[F] uncancelable { poll =>
+                      // invalidate the cancel action when we're done
+                      poll(fe.guarantee(Sync[F].delay(stateR.set(RegState.Completed))))
```
I think there could be a problem here: if `fe` finishes very quickly, this will unconditionally set the state to completed, even if there is a `CancelRequested` already there. If the `Worker` checks the state _after_ this, it won't find the `CancelRequested`, and this way the cancel will hang (the promise will never be completed).

At least I _think_ that's what happens. I have a branch, where there is a test which (sometimes) hangs; also, in the next commit, I'm trying to fix it: https://github.com/durban/cats-effect/tree/dispatcher-bug. (Seems to work, or at least the test doesn't hang any more.)

Feel free to merge my branch into this one if you think it's useful.

djspiewak replied:
Hmmmmmmmmmmmmmmmmmmmmmmmmm! I'll take a look! Thank you!

djspiewak replied:
Ah I agree this is a problem. I stared at it for a little while before I got it, but the issue manifests when the action completes before the `execute` lambda has a chance to run, meaning that the state is still `Unstarted`. So the exact sequencing:

1. We start running the action in `execute` (`stateR == Unstarted`)
2. We cancel the action in `cancel()` (`stateR == CancelRequested(latch)`)
3. The action completes and we unconditionally flip the state (`stateR == Completed`)
4. The `execute` lambda finally has a chance to run. The CAS fails but the state is `Completed` so we fall through to `Applicative[F].unit`
5. *`latch` is still waiting*

I think the solution is to `getAndSet` and complete the latch if present.

durban replied:
Yes, I think that's what I did on my branch (if I understand you correctly).

djspiewak replied:
You do and you did! I merged your changes. Interestingly, I can't get your test to *ever* fail on my M2 Max. My guess is the hardware timing is such that it's too hard to reproduce that narrow timeslice. Very nice catch.

durban replied:
Yeah, when I was playing with it, I noticed that the test was _really_ sensitive to _very_ small things.

durban commented on std/shared/src/main/scala/cats/effect/std/Dispatcher.scala:
```diff
@@ -190,240 +198,456 @@ object Dispatcher {
-                    }(ec)
+                } else {
+                  val withCompletion =
+                    action.guarantee(Sync[F].delay(reg.stateR.set(RegState.Completed)))
```
Also I think, this is redundant (we're already setting the state to completed above). (This is also removed in my branch.)

durban reviewed (commented):
I've left a few comments. ([One of them](https://github.com/typelevel/cats-effect/pull/3923#discussion_r1470187188) is I think about a bug and a fix.)

djspiewak committed:
Update std/shared/src/main/scala/cats/effect/std/Dispatcher.scala

Co-authored-by: Daniel Urban <urban.dani@gmail.com>

djspiewak committed:
Update std/shared/src/main/scala/cats/effect/std/Dispatcher.scala

Co-authored-by: Daniel Urban <urban.dani@gmail.com>

djspiewak committed:
Merge branch 'series/3.5.x' into feature/new-dispatcher

djspiewak committed:
Added missing import

djspiewak committed:
Merge branch 'series/3.5.x' into feature/new-dispatcher

armanbilge committed:
Fix `UnsafeAsyncQueue`

djspiewak committed:
Merge remote-tracking branch 'durban/dispatcher-bug' into feature/new-dispatcher

djspiewak committed:
Fixed gadt issues on 2.12

djspiewak commented:
Hold on this until we can track down the uncaught supervisor state exceptions.

durban committed:
Handle stateR more precisely

durban commented:
I believe I've fixed _one_ of the sources of the uncaught exceptions: #3991 (feel free to merge that branch into this one). The test affected was `complete / cancel race`. There was a race with multiple completions of a `CancelRequested(latch)` (with `trySuccess`), and one of these would also submit a `Fiber#cancel` to the cancelling-supervisor. But others would've already completed the `latch`, so the cancel-future would be completed, so the test could finish, so the supervisor could shut down. The fix in #3991 is to handle the state properly, and make sure there is only one completion of the latch.

There remain other sources of the same exception:

- `run multiple IOs in parallel`: I don't really understand this test; it seems to intentionally shut down a dispatcher while things are still going on?
- `cancelation race does not block a worker`: I think (but I'm not sure) this causes a `Worker` to finalize itself (with enqueing `PoisonPill`), then right after that the cancelling-supervisor finalizes itself (because it's the next `Resource` to close), but the queue of the worker is not empty, there could still be `Registration.Finalizer`s in it, which are then submitted to the cancelling-supervisor.

durban commented:
For the sequential, non-cancelable dispatcher, is `cancel()` supposed to be a no-op? Or is the future returned by `cancel()` supposed to be completed only once the task is finished?

Because currently (i.e., in this PR) it nondeterministically behaves either way, based on whether `cancel()` wins or loses a race. I.e., the following test sometimes hangs, sometimes completes for `Dispatcher.sequential[IO](await = true)`:

```scala
      "foobar" in real {
        D.use { dispatcher =>
          IO.fromFuture(IO {
            val (_, cancel) = dispatcher.unsafeToFutureCancelable(IO.never[Unit])
            val cancelFut = cancel()
            cancelFut
          })
        }.replicateA_(6).as(ok)
      }
```

djspiewak commented:
> For the sequential, non-cancelable dispatcher, is cancel() supposed to be a no-op? Or is the future returned by cancel() supposed to be completed only once the task is finished?

I think the latter is the behavior I would expect. `cancel()` should basically behave as if the whole task is wrapped in `uncancelable`. Nondeterministically behaving in one way or the other sounds like a gloriously weird bug.

Also I'm chewing on your branch right now. Will merge and then get to work on some of the other bits.

djspiewak committed:
Avoid weirdly depending on cancelation semantics in parallel deadlock test

djspiewak commented:
I fixed the weird parallel IO test to avoid dependence on await/non-await semantics. I still see unhandled exceptions though. It's really difficult for me to reproduce those exceptions on this machine. Also because of parallel specs execution it's very possible the exceptions are coming from somewhere other than those tests, but I'm not sure.

**Edit:** Actually now that I think about it, the nondeterminism in sequential `cancel()` *is* consistent with what would happen if the action was just wrapped in `uncancelable`. Consider:

```scala
IO.never.uncancelable.start.flatMap(_.cancel)
```

This will nondeterministically hang depending on whether the `cancel` gets in before the fiber begins running. Same idea here. So I *think* it's kind of okay?

djspiewak committed:
Unconditionally eat supervisor errors on finalizers

durban commented:
> This will nondeterministically hang depending on whether the cancel gets in before the fiber begins running. Same idea here. So I think it's kind of okay?

Except in the case of `start`/`cancel`, the uncancelable task either (1) doesn't even start executing (and `cancel` finishes quickly), or (2) executes "completely", i.e., hangs in `never` (and `cancel` also hangs).

With this dispatcher (note, we're still talking about sequential, cancelable = false, await = true) I think there is a (3): the task starts executing, but `cancel` finishes before the task is completed.

This test fails:

```scala
      "foobar2" in real {
        D.use { dispatcher =>
          IO.ref(0).flatMap { ctr1 =>
            IO.ref(0).flatMap { ctr2 =>
              IO.fromFuture(IO {
                val (_, cancel) = dispatcher.unsafeToFutureCancelable(IO.uncancelable { _ =>
                  ctr1.update(_ + 1) *> IO.sleep(0.1.second) *> ctr2.update(_ + 1)
                })
                val cancelFut = cancel()
                cancelFut
              }).flatMap { _ =>
                // if we're here, `cancel()` finished, so
                // either the task didn't run at all (i.e.,
                // it was cancelled before starting), or
                // it ran and already finished completely:
                (ctr1.get, ctr2.get).flatMapN { (v1, v2) =>
                  IO(v1 mustEqual v2)
                }
              }
            }
          }
        }.replicateA_(10000).as(ok)
      }
```

djspiewak commented:
Ah, okay I buy that this is a problem then.

djspiewak commented:
Minor update: I've seen this test fail once, but it's very difficult to get it to fail at all on my laptop. Something about the timing I suspect. I may need some help tracking this one down.

durban committed:
Add failing test

durban committed:
Fix it

durban committed:
scalafmt

durban committed:
Only replicateA_ on JVM

durban commented:
I think I've fixed it. See #4012 (feel free to merge my branch directly into this PR).

The problem is with this comment:

```scala
// we can use unit as a cancel action here since it must always sequence *after* the task
// thus, the task must complete before the cancel action will be picked up
```

This is not (always) true:

1. It _is_ true if `registerCancel` in `Worker` modifies the state `Unstarted` -> `Running(pure(()))`. In that case, `cancel()` will find the `pure(())` in `Running`, and submit it to the _same queue_.
2. However, if `cancel()` is faster, changes state `Unstarted` -> `CancelRequested`. Then `registerCancel` in `Worker` changes state `CancelRequested` -> `Running(pure(()))`, then submits the `pure(())` to _the separate cancel-supervisor_. The cancel-supervisor "executes" the `pure(())`, then completes the `latch` in `CancelRequested` (`cancel()` is waiting for this). What the cancel-supervisor does, is (or can be) parallel to `Executor` running the orignal task.

My fix (in [6a9cfa6](https://github.com/typelevel/cats-effect/pull/4012/commits/6a9cfa617676887dd34ae450aff05dac0ddb5072)) is to instead of the synthetic `pure(())` cancellation action, wait for a `Deferred` for the task to complete. (Since we're ignoring cancellation, this should be fine in either case.) Without the fix, I could reliably reproduce the bug with the test in [a3c285d](https://github.com/typelevel/cats-effect/pull/4012/commits/a3c285dcbe9134394efe71d70957949d23de18ac). With the fix, I can't any more.

djspiewak commented:
Magnificent fix @durban! Sorry I had been too busy to get to this. Snagged it now. Thank you!

djspiewak committed:
Merge branch 'series/3.5.x' into feature/new-dispatcher

djspiewak commented:
Let's call this good. I got the 👍 from @armanbilge just now to merge without waiting for him, since @durban in particular has been all up in this PR's business. If nothing else, this should make the build a lot more stable!
